### PR TITLE
[@Powerpages Copilot] Context Improvement for Webpage

### PR DESCRIPTION
**Pull request by amitjoshi438**
This pull request introduces several enhancements to the `PowerPagesChatParticipant` class and related utilities, focusing on managing and fetching related files for a given context. The most important changes include adding support for fetching related files based on the active file's context, updating schemas for file extensions, and modifying API requests to include related files.

Enhancements to file management:

* [`src/common/utilities/Utils.ts`](diffhunk://#diff-bf8425ddaf7889e68044eb87abc71b32754d10ce9a306723063cf231b01f590cR223-R294): Added functions `getFileContent`, `getFileContentByType`, and `fetchRelatedFiles` to manage and fetch related files based on the active file's context.

Schema updates:

* [`src/common/constants.ts`](diffhunk://#diff-118a1ed2ebf5b946c94c5dfeae1c4d13c9a57ef105b36ad323e0b7b8e368da27R41-R65): Defined `componentTypeSchema` and `relatedFilesSchema` to map file types and their extensions, and added an interface `IRelatedFiles` for related files.

API request modifications:

* [`src/common/copilot/IntelligenceApiService.ts`](diffhunk://#diff-a61cdf8af8b4c524990ec9dfb07752b80f4e7dad3fae3a337383fe151d79c176L13-R19): Updated `sendApiRequest` to include an optional `RelatedFiles` parameter and modified the request payload to include related files. [[1]](diffhunk://#diff-a61cdf8af8b4c524990ec9dfb07752b80f4e7dad3fae3a337383fe151d79c176L13-R19) [[2]](diffhunk://#diff-a61cdf8af8b4c524990ec9dfb07752b80f4e7dad3fae3a337383fe151d79c176R43)

Class updates:

* [`src/common/chat-participants/powerpages/PowerPagesChatParticipant.ts`](diffhunk://#diff-885f58fd5bd7ee590ddae0f7854d056026aff9b576864e07f61887fb7586da29L18-R22): Integrated the `fetchRelatedFiles` function to fetch and include related files in the API request within the `PowerPagesChatParticipant` class. [[1]](diffhunk://#diff-885f58fd5bd7ee590ddae0f7854d056026aff9b576864e07f61887fb7586da29L18-R22) [[2]](diffhunk://#diff-885f58fd5bd7ee590ddae0f7854d056026aff9b576864e07f61887fb7586da29R163-R179)

Minor updates:

* [`src/common/copilot/constants.ts`](diffhunk://#diff-6205a2cccf0d15c5083495892549bfb3f44d49b232c6f2ce923277f02d239cebL66-R68): Added `css` to the `FieldTypeMap`.
* Removed an extraneous line in `src/common/chat-participants/powerpages/PowerPagesChatParticipantUtils.ts`.

### CODE REVIEW

File: src/common/chat-participants/powerpages/PowerPagesChatParticipant.ts [hunk 1]
```diff
@@ -159,9 +160,23 @@ export class PowerPagesChatParticipant {
...
+                    }
+                    break;
+                default:
+                    break;
+            }
+
             const { componentInfo, entityName }: IComponentInfo = await getComponentInfo(this.telemetry, this.orgUrl, activeFileParams, this.powerPagesAgentSessionId);
 
-            const llmResponse = await sendApiRequest([{ displayText: userPrompt, code: activeFileContent }], activeFileParams, this.orgID, intelligenceApiToken, this.powerPagesAgentSessionId, entityName, componentInfo, this.telemetry, intelligenceAPIEndpointInfo.intelligenceEndpoint, intelligenceAPIEndpointInfo.geoName, intelligenceAPIEndpointInfo.crossGeoDataMovementEnabledPPACFlag);
+            const llmResponse = await sendApiRequest([{ displayText: userPrompt, code: activeFileContent }], activeFileParams, this.orgID, intelligenceApiToken, this.powerPagesAgentSessionId, entityName, componentInfo, this.telemetry, intelligenceAPIEndpointInfo.intelligenceEndpoint, intelligenceAPIEndpointInfo.geoName, intelligenceAPIEndpointInfo.crossGeoDataMovementEnabledPPACFlag, relatedFiles);
```
Comment (tyaginidhi) [inline src/common/chat-participants/powerpages/PowerPagesChatParticipant.ts:179]: we should convert these param into an object now for better readability and param passing

File: src/common/constants.ts [hunk 2]
```diff
@@ -38,3 +38,30 @@ export const PORTAL_YEOMAN_GENERATOR_PACKAGE_TARBALL_NAME = "microsoft-generator
...
+export const componentTypeSchema: { [key: string]: { [key: string]: string } } = {
+    'adx_webpage': {
+        'html': '.copy.html',
+        'js': '.custom_javascript.js',
+        'css': '.custom_css.css'
+    }
+};
+
+// Define the schema
+export const relatedFilesSchema: { [key: string]: { [key: string]: string[] } } = {
```
Comment (tyaginidhi) [inline src/common/constants.ts:52]: not in this PR - but we should consider porting portalschema to common and make these mappings central for desktop and web extesion

File: src/common/utilities/Utils.ts [hunk 3]
```diff
@@ -219,3 +219,84 @@ export function getVisibleCode(editor: vscode.TextEditor): { code: string; start
...
+// Generic function to get file content based on type and component type
+async function getFileContentByType(activeFileUri: vscode.Uri, componentType: string, fileType: string): Promise<{ customFileContent: string; customFileName: string; }> {
+    try {
+        const extension = componentTypeSchema[componentType]?.[fileType];
+        if (!extension) {
+            throw new Error(`File type ${fileType} not found for component type ${componentType}`);
+        }
+        return await getFileContent(activeFileUri, extension);
+    } catch (error) {
+        const message = (error as Error)?.message;
```
Comment (tyaginidhi) [inline src/common/utilities/Utils.ts:260]: what experiene will user have in case of error like this?

File: src/common/utilities/Utils.ts [hunk 4]
```diff
@@ -219,3 +219,84 @@ export function getVisibleCode(editor: vscode.TextEditor): { code: string; start
...
+
+//fetchRelatedFiles function based on component type
+export async function fetchRelatedFiles(activeFileUri: vscode.Uri, componentType: string, fieldType: string, telemetry: ITelemetry, sessionId:string): Promise<IRelatedFiles[]> {
+    try {
+        const relatedFileTypes = relatedFilesSchema[componentType]?.[fieldType];
+        if (!relatedFileTypes) {
+            return [];
+        }
+
+        const files: IRelatedFiles[] = await Promise.all(
```
Comment (tyaginidhi) [inline src/common/utilities/Utils.ts:273]: be careful on how you are using and expecting promise.all to do here: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/all

File: src/common/utilities/Utils.ts [hunk 5]
```diff
@@ -219,3 +219,84 @@ export function getVisibleCode(editor: vscode.TextEditor): { code: string; start
...
+                    fileContent: fileContentResult.customFileContent,
+                    fileName: fileContentResult.customFileName
+                };
+            })
+        );
+
+        return files;
+    } catch (error) {
+        const message = (error as Error)?.message;
+        telemetry.sendTelemetryErrorEvent(COPILOT_RELATED_FILES_FETCH_FAILED, { error: message, sessionId: sessionId });
```
Comment (tyaginidhi) [inline src/common/utilities/Utils.ts:287]: check user experience on this type of failures

**Review by tyaginidhi — APPROVED** [review]

File: src/common/utilities/Utils.ts (hunk 3 above)
Comment (amitjoshi438) [reply src/common/utilities/Utils.ts:260]: User's will not see this error, the related files will not be passed to LLM and there will be variation in response.

File: src/common/utilities/Utils.ts (hunk 4 above)
Comment (amitjoshi438) [reply src/common/utilities/Utils.ts:273]: Yes, I believe you are referring to rejection behavior of promise.all(). Currently we want to pass related files only when all the related files are present.

File: src/common/chat-participants/powerpages/PowerPagesChatParticipant.ts (hunk 1 above)
Comment (amitjoshi438) [reply src/common/chat-participants/powerpages/PowerPagesChatParticipant.ts:179]: Right, have already made that change in next PR.

**Review by gshivi — APPROVED** [review]

**Pull request merged**